### ISSUE TITLE
Use type classes for input summands

### ISSUE THREAD

**Issue by paf31** [slamdata/purescript-halogen#13]
Things like undo/redo and error handling require the input to be a sum (`Either Error`, `UndoRedoInput` etc.)

Using type classes to represent these "input mixins" would be preferable to requiring specific types because otherwise, we have to use the `profunctor` combinators to get the `Either`s in the right order to pass to `runUIAff`.  

E.g.

```purescript
class HasUndoRedo input where
  undo :: input
  redo :: input
  toUndoRedo :: input -> Maybe UndoRedoInput

class HasError input where
  error :: Error
  toError :: input -> Maybe Error
```

**Comment by jdegoes** [comment]
I was thinking of this approach for my work with channels. Another approach that occurred to me is abusing row types:

```purescript
(hasUndoRedo :: HasUndoRedo input | e)
```

Both of them amount to a product encoding of sum types (i.e. big old product of maybes).

I'm not sure which would be less boilerplate in the long run.

**Comment by paf31** [comment]
Clearly we need polymorphic variants :(

**Comment by jdegoes** [comment]
That'd be pretty sweet. :smile:

**Comment by paf31** [comment]
I implemented them sometime around 0.2 in a branch, but there was an issue: they don't play nicely with constructors-as-functions. There's an unfortunate asymmetry between sums and products here: every field in a record has one arguments, but every data constructor in a sum takes zero or more arguments. So if you say

```purescript
\x -> Just x
```

there is no principal type any more. It could be

```purescript
forall rest a. a -> Sum (Just :: [a] | rest)
```

or 

```purescript
forall rest a b. a -> b -> Sum (Just :: [a, b] | rest)
```

or 

```purescript
forall rest a b. a -> b -> c -> Sum (Just :: [a, b, c] | rest)
```

etc.

**Comment by jdegoes** [comment]
@paf31 It would probably introduce other problems, but if one re-imagines terms in a sum type as being data types (i.e. both constructors and types), then this particular issue goes away.

e.g. every sum type becomes:

```purescript
type Maybe a = Just a | Nothing
data Just a = a
foreign import data Nothing :: *
```

I guess that changes them to union types instead of sum types, and who knows what effects that would have on the rest of the type system. Oh well!
